### ISSUE TITLE
Weird rendering bug on Windows only

### ISSUE THREAD

**Issue by mfncooper** [urwid/urwid#774]
Using a full-screen LineBox as the main window area, I am seeing a weird character in the left-most column of every row. I don't know what the character is - it looks like an outline of a square with tiny lines coming out of top, bottom, left and right. And because of this character in the left-most column, everything else gets shifted off to the right by one character, so I lose the right-most column. There is also one instance of the same character at the bottom right of the screen.

This only happens on Windows. (I am using Windows 10 with Python 3.10.11.) The same code works perfectly on both Mac and Linux. In all cases, I'm using Urwid 2.4.6 installed from PyPI using pip.

I've stripped down the code to a short app that reproduces the problem, and included that here. I've also included a screenshot of the app when running under Windows.

```python
import urwid

class ContentPanel(urwid.WidgetWrap):
    def __init__(self):
        self._text = urwid.Text("Hello, World!")
        self._fill = urwid.Filler(self._text)
        super().__init__(self._fill)

class MainScreen(urwid.WidgetWrap):
    def __init__(self, panel):
        pile = urwid.Pile([('weight', 1, panel)])
        super().__init__(urwid.LineBox(
            pile, title="Main Screen", title_align='center'))

def create_widgets():
    content_panel = ContentPanel()
    main_screen = MainScreen(content_panel)
    frame = urwid.Frame(main_screen)
    return frame

def unhandled_input(key):
    if key == 'q' or key == 'Q':
        raise urwid.ExitMainLoop()
    return False

def run():
    loop = urwid.MainLoop(
        create_widgets(),
        unhandled_input=unhandled_input)
    loop.run()

if __name__ == "__main__":
    run()
```

![RenderBug](https://github.com/urwid/urwid/assets/450625/c2631cdd-f62c-45f7-9f16-a3e71ec22093)

**Comment by penguinolog** [comment]
Maybe Unicode processing issue: consistent Unicode symbol on screenshot.
Need to validate, that "stupid" `print(create_widgets().render((80, 24)))` is correct.
Expect that `print` will work correctly and issue is somewhere in final screen rendering (windows is "special").

**Comment by mfncooper** [comment]
@penguinolog You are correct. I replaced `run()` in the above code with `print(create_widgets().render((80, 24)))` as you suggested, and the output was correct. I also tried writing the `print()` output to a file (as UTF-8, of course) to verify that there are no "invisible" characters, and there are none.

I'd love to get this resolved. It's the only issue that prevents my soon-to-be-released app from running on Windows; everything else works on all platforms. If there's any way I can help, let me know. I'm not afraid of digging into Urwid code if someone can direct me. I've been in it before, but not extensively, and not for a while. Thanks.

**Comment by mfncooper** [comment]
By changing the font to 72pt, I was able to see the weird character better, and have included a fragment of the image here. This looks like the Unicode character named "Gear Without Hub", code U+26ED. Maybe that can help track this down.

![image](https://github.com/urwid/urwid/assets/450625/bfe9aaf4-521f-49bf-8751-e26c9e60fa00)

**Comment by mfncooper** [comment]
Another update. I've found that it is unrelated to either `LineBox` or `WidgetWrap`. First I tried changing `create_widgets()` to just return `content_panel`, which is just a `Text` wrapped in a `Filler`, and got the same result. Then I eliminated the `WidgetWrap` subclasses so that `create_widgets()` was only the following, and the problem persists.

```python
def create_widgets():
    text = urwid.Text("Hello, World!")
    fill = urwid.Filler(self._text)
    return fill
```

**Comment by penguinolog** [comment]
Please try with patch from linked PR. It's relevant to the similar render issues on WSL, where SI/SO sequences generated garbage in the output

**Comment by penguinolog** [comment]
Issue #264 looks the same

**Comment by mfncooper** [comment]
Heh. I was going down the same path, and had discovered that SI was causing the weird characters. Yes indeed, 2.5.1 (which includes the referenced PR) fixes the weird character issue. Thank you! And thanks for the super-fast turnaround on releases like this.

One strange thing that I did not expect is that installing 2.5.1 pulls in the typing_extensions package. Is that really needed now?

There is one more issue in the rendering of my app, related to escape code differences with Windows, but not directly related to the weird characters. I will create a separate issue for that. I also have a (very simple) proposed fix for it.

**Comment by penguinolog** [comment]
Typing extension are really used, because `typing.Protocol` is not available for python 3.7 and protocol class used un runtime check.
At this moment the main python versions for which urwid is downloaded: 3.7 and 3.10.
When 3.7 usage decline to the levels of current 3.8 (small amount of downloads), we can deprecated it and re-review compatibility requirements

**Comment by mfncooper** [comment]
Well, the thing is that the dependency is **new**. I don't know if it appeared in 2.5.0 or 2.5.1, but there was *no* dependency in 2.4.6, so this is the first time Urwid has ever pulled in *any* other package as a dependency. This isn't just a development dependency?

**Comment by penguinolog** [comment]
> Well, the thing is that the dependency is **new**. I don't know if it appeared in 2.5.0 or 2.5.1, but there was *no* dependency in 2.4.6, so this is the first time Urwid has ever pulled in *any* other package as a dependency. This isn't just a development dependency?

It's runtime dependency in 2 places

**Comment by mfncooper** [comment]
Disappointing that Urwid now has a dependency.

In any case, I'll close this as fixed, since v2.5.1 does solve the weird character issue. Thanks!

**Comment by penguinolog** [comment]
Urwid will start using `wcwidth` as dependency in the future: it's more
efficient to use almost "world standard" instead of maintenance of private
the same functionality. This will fix a lot of problems with wide
characters (but not all).

Op vr 2 feb 2024 00:02 schreef Martin Cooper ***@***.***>:
